### PR TITLE
cuda_abort: remove [[no return]] with KOKKOS_ENABLE_DEBUG_BOUNDS_CHECK

### PR DESCRIPTION
**Pull request by ndellingwood**
addresses cuda random number generator test failures on Pre-Volta archs

### CODE REVIEW

**Comment by ndellingwood** [comment]
These changes address issues with random number generator test failures on Pre-Volta GPUs with `KOKKOS_ENABLE_DEBUG_BOUNDS_CHECK` set, detected during integration testing for the 3.2 release. This also resolved issues in graph algorithm unit test in kokkos/kokkos-kernels#785

Tested on the Ride testbed (Power8, Pascal60) with the following configuration:

```bash
module load cuda/9.2.88 cmake/3.12.3 gcc/7.2.0 git

cmake -DCMAKE_CXX_COMPILER=${KOKKOS_PATH}/bin/nvcc_wrapper -DCMAKE_MAKE_PROGRAM=make -DCMAKE_INSTALL_PREFIX="${KOKKOS_INSTALL_PATH}" -DKokkos_ENABLE_CUDA=ON -DKokkos_ARCH_PASCAL60=ON -DKokkos_ARCH_POWER8=ON -DKokkos_ENABLE_TESTS=ON -DKokkos_ENABLE_EXAMPLES=OFF -DKokkos_ENABLE_CUDA_LAMBDA=ON -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_CXX_EXTENSIONS=OFF -DCMAKE_BUILD_TYPE=DEBUG -DKokkos_ENABLE_DEBUG_BOUNDS_CHECK=ON -DKokkos_ENABLE_DEBUG=ON ${KOKKOS_PATH}
```

**Comment by ndellingwood** [comment]
Cross referencing #3106

File: core/src/Cuda/Kokkos_Cuda_abort.hpp [hunk 1]
```diff
@@ -65,7 +65,11 @@ namespace Kokkos {
 namespace Impl {
 
 #if !defined(__APPLE__)
+#if defined(KOKKOS_ENABLE_DEBUG_BOUNDS_CHECK)
```
Comment (dalg24) [inline core/src/Cuda/Kokkos_Cuda_abort.hpp:68]: Please comment above this line that this was causing issues with the random number generator unit tests on pre-Volta GPU architectures.

File: core/src/Cuda/Kokkos_Cuda_abort.hpp [hunk 2]
```diff
@@ -74,8 +80,12 @@ namespace Impl {
   // This loop is never executed. It's intended to suppress warnings that the
   // function returns, even though it does not. This is necessary because
   // __assertfail is not marked as [[noreturn]], even though it does not return.
+  //  Disable with KOKKOS_ENABLE_DEBUG_BOUNDS_CHECK to workaround failures
+  //  in random number generator unit tests with pre-volta architectures
+#if !defined(KOKKOS_ENABLE_DEBUG_BOUNDS_CHECK)
```
Comment (ndellingwood) [inline core/src/Cuda/Kokkos_Cuda_abort.hpp:85]: Guarding out the `while` loop here was necessary to resolve the unit test failures with the rand num gen; the other changes helped suppress warnings

**Comment by ndellingwood** [comment]
Cherry-picked to release-candidate-3.2 via
`git cherry-pick -m 1 -x 126f8429d9f7b10c8d218f4329a4bcebff7544d8`

**Pull request merged**